### PR TITLE
Added support of IN-statement when a list is passed to conditions as the value.

### PR DESCRIPTION
**Pull request by romul**
```elixir
{query, params} = Query.select(:users, [a: [1, 5, 7], b: 2])
 
assert query == "SELECT * FROM users  WHERE a IN ($1, $2, $3) AND b = $4"
assert params == [1, 5, 7, 2]
```

### CODE REVIEW

File: lib/defql/adapter/postgres/query.ex
Comment (implicitly-awesome) [inline lib/defql/adapter/postgres/query.ex]: Mb it could be done with `Enum.map_join/3` without overrun? :)

Comment (romul) [reply lib/defql/adapter/postgres/query.ex]: Good point! Fixed: https://github.com/fazibear/defql/pull/1/commits/5af9aa4784418a27907fc22f91eb1da58394152b

**Review by implicitly-awesome — APPROVED** [review]

**Comment by implicitly-awesome** [comment]
👍

**Comment by fazibear** [comment]
Thanks for the contribution! Firstly I want to notice that I was thinking about something more generic. Because array can be also used as array type while inserting data, it can be a little bit confusing. 

Maybe something like this:
```
Query.select(:users, a: {:in, [1, 5, 7]}, b: 2)
```
so we can implement other keywords like `like` or `ilike`. We should consider that.
Also, same syntax should be used in `defupdate` macro, there also is a condition part.

Another thing is a code. I really don't like `do_get_conditions` function. The name is ugly tells nothing about it. Also, recurrence looks little messy.

We should consider both things and  refactor this code a little.
What do you think ?

**Comment by romul** [comment]
Hmm, I personally like this popular naming convention with `do_` prefix for a private part of the function. It's pretty obvious that `do_` functions should not be called directly even inside their module (the only exception is a wrapper-function without `do_` prefix). 
But we could use `get_conditions` for all clauses like that:
```elixir
  defp get_conditions(params, idx \\ 0)
  defp get_conditions([], _), do: ""
  defp get_conditions(params, idx) do
    get_conditions(params, idx+1, [])
  end

  defp get_conditions([], _, acc) do
    [" WHERE ", acc |> Enum.reverse |> Enum.join(" AND ")]
  end
  defp get_conditions([{field, value} | other_conds], idx, acc) when is_list(value) do
    count = Enum.count(value)
    placeholders = (idx..idx+count-1) |> Enum.map_join(", ", &("$#{&1}"))
    condition = "#{field} IN (#{placeholders})"
    get_conditions(other_conds, idx + count, [condition | acc])
  end
  defp get_conditions([{field, _} | other_conds], idx, acc) do
    get_conditions(other_conds, idx + 1, ["#{field} = $#{idx}" | acc])
  end
```


As to syntax, for `like` or `ilike` it looks good, but for `in` it's not so intuitive. Although introducing of this syntax with my solution would be very trivial. I'll commit it in a few minutes.

**Comment by romul** [comment]
Done.

**Comment by fazibear** [comment]
Can you take a look at my implementation? https://github.com/fazibear/defql/pull/4
Don't you think that this one-way data flow implementation is a way cleaner and easier to follow?

**Comment by romul** [comment]
@fazibear Honestly, I don't see *your implementation*, I see my implementation with a few very minor changes.. like use of `get_indicies` instead of optimized solution, replacing explicit recursion with Enum.map_reduce/3 and some renaming.
So you could just add some commits according to your personal preferences on top of my commits (without changing of authorship of my code), for example:  https://github.com/fazibear/defql/pull/1/commits/a624f138fbf96bc3ed91a16b643cf1e5bab958ec and https://github.com/fazibear/defql/pull/1/commits/978ee437d7698449dafedf823e33cb288bfeb1fb.

> Don't you think that this one-way data flow implementation is a way cleaner and easier to follow?

Hmm, I didn't get you. The data flow in your PR is totally the same as in mine.

**Comment by fazibear** [comment]
Things I don't like about your implementation:
- recursion - hard to follow `get_conditions` flow, you need to check params
- `get_conditions` with accumulator
- `get_conditions` with reverse statement
- enumerate with ranges
- a lot of assignments

Seriously, I tried to modify your code, but it was easier to implement it from scratch.

**Comment by fazibear** [comment]
I really want to keep things simple!

**Comment by romul** [comment]
@fazibear I updated my PR a few hours ago and implemented these suggestions.
I totally recommend you to learn recursion, if such a simple and common recursion techniques is difficult to follow for you. `Enum.map_reduce/3` is not a magic and it works almost exactly as my previous code. As you could see from https://github.com/fazibear/defql/pull/1/commits/978ee437d7698449dafedf823e33cb288bfeb1fb change was quite trivial.
But you pretended not to notice new commits in my PR and prefer just to steal my solution by committing it under your name. I don't think, that anybody else would like to help you after such the incident.

**Comment by fazibear** [comment]
@romul Please don't judge my knowledge like that. I know what recursion is and how to use it. But in this case was not needed in my opinion. So I've started to write my code without it. That's it. Also, I think that implementation without recursion is more readable.

If you're still thinking that code was stolen, please show me exactly what part? Because your solution is base on recursion, mine not. Your code has a lot of assignments, mine not. If you just look at diffs, the code is completely different, different techniques, different function names, the different solution, different code style.

I want to keep code clean and easy to read, it's nothing wrong with that. That's why I've chosen other solution.

It's really hard to believe that you accuse me of stealing.

**Comment by romul** [comment]
@fazibear 

> But in this case was not needed in my opinion.

map_reduce does exact the same recursion, so it's definitely needed in this case. I have agreed with replacing explicit implementation with call of map_reduce and done it in my PR before you have merged yours commits. And again the implementation remains exactly the same, the only difference is changing order of arguments in one function.

> If you're still thinking that code was stolen, please show me exactly what part? 

Just compare:

https://github.com/romul/defql/blob/in_statement_support/test/defql/adapter/postgres/query_test.exs#L19..L45
https://github.com/fazibear/defql/blob/master/test/defql/adapter/postgres/query_test.exs#L19..L45

https://github.com/romul/defql/blob/in_statement_support/lib/defql/adapter/postgres/query.ex#L79..L93
https://github.com/fazibear/defql/blob/master/lib/defql/adapter/postgres/query.ex#L80..L91

https://github.com/romul/defql/blob/in_statement_support/lib/defql/adapter/postgres/query.ex#L107..L131
https://github.com/fazibear/defql/blob/master/lib/defql/adapter/postgres/query.ex#L104..L120

Yes, you slightly changed my code, but there are no significant differences, it's exactly the same solution. The only visible difference between your code and my code is naming and minor differences in code style. I guess you would be happy if I rename your project and release it under my name? It's silly.
You are definitely kidding me trying to say that this is not a plagiarism. At least, in polite society people do not behave like you. Such shameless code authorship changes I have not seen even once in the 7 years of active contribution in OSS.

**Comment by fazibear** [comment]
> map_reduce does exact the same recursion

Recursion happened when the function calls itself. There is no case like that. map_reduce iterate over the collection with accumulator which is different. 

OK, Because things are getting really serious. 

Just like I said, I didn't take any of your ideas. All code was written from scratch on master and pushed into another branch. I did it because didn't like a way that your code was written. And there is no sense to rewrite it anyway. Just like I told you, the code is completely different.

The only thing that was copied is a part of the test case, just to show you that my code has exactly the same output. Sorry about that, I was focused on implementation.

And let me repeat myself. In my version there is no recursion, no case statement, new code style matches old one. My code is not even inspired by yours because I don't like it. 

As we can see in the log, you've changed you code during the day, and your changes was inspired by my implementation, but I don't think you steal something from me. 

I just don't like your pull request, and it's nothing weird. It can happen.

I should just close your PR and forget about it, but I want to know your opinion about my implementation, what I get is accuse me of stealing. 

If you want I'll remove all code that you think is similar to yours. I just want to close this case.

**Comment by romul** [comment]
Let's ask opinion of independent reviewer @madeinussr...

> I want to know your opinion about my implementation

Yep, you have started discussion, waiting for my answer and just ignore it by committing "your" version.

> As we can see in the log, you've changed you code during the day, and your changes was inspired by my implementation

My changes were motivated by your suggestions. Once again they all are about styling, the solution is the same since the first commit. And that's ok when repo owner ask for changes in PR, like [here](https://github.com/romul/newrelic.ex/pull/3), but it's not ok to slightly change another's code and commit it as yours.

> If you want I'll remove all code that you think is similar to yours. 

By another renaming and "refactoring"? It's silly. The only way to save your blushes is to restore authorship of my commits. In any case, I don't want to continue this discussion. Just disgusting to know that people could behave like you in the OSS community.

**Comment by fazibear** [comment]
As you know I didn't merge your PR because I didn't like implementation. I'm tried to go thru your implementation, but a better way to show you what I mean is to implement it by myself. It makes no sense to merge PR delete all changes and write it once again, so another PR was merged. BTW, I have the implementation of that feature in my mind for a long time. When I've started write whole module. That's why it fits better.

I really don't want to waste my time to convince you that other people can implement the same thing in another way without using someone code, not even looking at it. Right now a lot of people are writing code similar to yours right now, and they are not stealing.

Just like I said. I'll remove all code that you're thinking is authored by you, and we will finish this case.

**Comment by fazibear** [comment]
@romul Sorry about this incident. I talked to some people what really happened here, and think that we don't understand each other. I really appreciate your work. I just make another PR and share it with you just to show you how I imagine the implementation of this feature. I made a few mistakes, used few inappropriate words. Don't want to make the same mistake again.

**Comment by implicitly-awesome** [comment]
As I was mentioned here, here is my opinion.
Definitely, there were a bunch of inappropriate words. 'Steal' is very strong and negative word, indeed. But, I understand @romul position too. There is not rocket-science task, so the most of developers could implement it in the same way (besides naming and either map/reduce or recursion usage. Although map/reduce are implemented through recursion itself and are just syntactic sugar).

_From my perspective, @fazibear could request some changes through comments to @romul implementation in the certain PR (you see, through discussion)._

So, I understand both @romul and @fazibear complains, one's complaining about time the implementation took, other's take care for the code organisation as the owner of the repo with his own code structuring vision.

PS: sorry for my English. I hope I explained my opinion clearly.

**Comment by romul** [comment]
@fazibear I admit that I overreacted... I can imagine that you have written a similar solution by yourself.. is not the point though. I just judge from from my own experience in the PR acceptance/declining. I have accepted hundreds of PRs. And even when I have changed another's code I just did `commit --amend`, because in my personal opinion it is the only acceptable way of working with the community (and the way to express appreciation for the assistance). So, I was very surprised by your reaction: "Hey, I can redo this task by myself". It's obvious that every developer can redo any task (especially if it's already done). But I agree with @madeinussr, any requested changes could be done through discussion & collaboration. It's the way how Open Source works.
After all, the changes have been requested and have even been made by me. 
You could just say "no thanks, I prefer to do everything by myself" after the first commit. This is a clear position. But when you at first request changes and after they are done simply ignore all the work - it is disrespectful.

**Comment by zoldar** [comment]
To be clear, what's written below is strictly my (humble) personal opinion.

As @madeinussr already stated, given a small and well defined idea, the range of sensible solutions to it is pretty limited. However, I believe that the point of @fazibear's rewrite was not to come up with diametrically different solution, but to simply make the implementation of that idea more consistent with the rest of the existing codebase. Also, he found it easier and less time consuming to express it in code than a series of suggestions to the existing PR.

Although the effort of the person making the contribution is definitely commendable, the weight of actually maintaining that code throughout the life of the project is on the shoulders of the maintainer. Added to that, there are only so many hours in a day and only a lucky few have the luxury of grooming the oss projects as part of their day job. Given the limited amount of time and energy that the maintainer can give, it's in the project's best interest to choose what's least time consuming for the maintainer. It's of course always desired to give proper guidance to contributors and recognize their work in the best way possible but when resources are scarce, choices have to be made.

This is not to say that the whole situation couldn't be handled better - perhaps, @fazibear could hold on with merging his own solution until he got at least a cursory feedback on it. Perhaps he could explicitly refer to the @romul's PR (either in the PR or in the commits) and make it clear that it's the initial proposal on the basis of which merged implementation got created. Perhaps (and that's something that still could be done) it would be a good idea to create a "Contributors" section in the project's `README` and put @romul there with reference to both PRs so that attribution would still be there (something along the lines of https://github.com/zoldar/python-ldap-test#contributors but with additional context; no shameless plug intended ;) )? All future contributors with reference to relevant PRs could be put there as well.

From a purely technical standpoint, even though I can follow both implementations of the idea (including the one using pure recursion), I think that using abstractions which are well established in the FP world (like `map`, `reduce` or their combination) increases the ease with which the codebase can be read. Those abstractions are well understood and remove the cognitive load of having to unwrap the recursion in one's head, identifying accumulator etc. It simply saves time and ultimately increases readability. But again, this is highly dependable on context and I can be biased ;)

Ok, that were my 3 cents. In the end, I think that it's simply not worth it to have bad blood with anybody over a thing as mundane as code. We are not our code. Also, I don't suspect anyone here of having bad intentions. Peace out :)

**Comment by romul** [comment]
To sum up, I think everybody will stay with own opinion. So, let's write code instead of continuation of this discussion.

**Pull request closed without merging**
